### PR TITLE
OCPBUGS-8447: MCO-496: Support ignition versions 3.3 + 3.4 but keep version 3.2 as default 

### PR DESCRIPTION
**Pull request by jkyros**
This allows the MCO to receive and serve ignition versions 3.3 and 3.4, but DOES NOT change the default from 3.2, and as discussed *does not* include MCO feature implementations for the kernel args feature nor the "offline tang" feature (lest we introduce more risk).

 The MCO has traditionally tried to support an ignition version at least one release before it's actually used so we have the ability to roll back an upgrade, etc. This PR is in service to that goal so we can use 3.4 in the next version, and next version we will come back and do the "rest of the work" to move the default version to 3.4. 

**- What I did**

Added ignition 3.3 version support (but not support for [the kargs feature yet](https://github.com/openshift/machine-config-operator/pull/2675)) 
Added ignition 3.4 version support (but not support for the offline tang feature yet) 
Updated unit tests to account for the new versions
Bumped ignition deps for ignition 3.4

This means that: 
- Ignition versions > 3.2 will be "down-translated" to 3.2 
- Ignition versions < 3.2 will be still be "up-translated" to 3.2
- If something asks the MCS for a 3.3 or 3.4 version, it will "up-translate" the 3.2 to that version to serve it

**- How to verify it**
Feed the MCO something in the ignition 3.3/3.4 format:

Feed the MCO something in the ignition 3.3/3.4 format that contains fields not supported in 3.2 (kargs, offline tang): 

Request a 3.3/3.4 version ignition from the MCS:

**- Description for the changelog**
Allow MCO to serve and receive ignition spec versions 3.3 and 3.4

### CODE REVIEW

File: pkg/controller/common/helpers.go [hunk 1]
```diff
@@ -519,11 +543,19 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
...
 func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
 	// ParseCompatibleVersion will parse any config <= N to version N
-	ignCfgV3, rptV3, errV3 := ign3_3.ParseCompatibleVersion(rawIgn)
+	ignCfgV3, rptV3, errV3 := ign3_4.ParseCompatibleVersion(rawIgn)
 	if errV3 == nil && !rptV3.IsFatal() {
+		// Regardless of the input version it has now been translated to a 3.4 config, downtranslate to 3.3 so we
+		// can get down to 3.2
+		ignCfgV3_3, errV3_3 := v34tov33.Translate(ignCfgV3)
+		if errV3_3 != nil {
+			// This case should only be hit if fields that only exist in v3_4 are being used which are not
```
[the remark this answers is not in the document]
Comment (jkyros) [reply pkg/controller/common/helpers.go:552]: Yep, correct. The "tang/kernel args fields" are exactly what I'm referring to. 

The checks you're describing do exist but not in the MCO -- they're in ign-converter, e.g;

https://github.com/coreos/ign-converter/blob/897284baf1179e76ac6ef14e8ca8542334d58c55/translate/v34tov33/v34tov33.go#L125

and they're applied as part of the "down-translation" process, so I'm "freeloading" on that so we don't have to duplicate that bit of code into the MCO.

And so yes,  if you feed us an ignition containing offline tang fields: 
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name:  99-cant-convert-34
spec:
  config:
    ignition:
      version: 3.4.0
    storage:
    luks:
    - name: luks-tang
      device: "/dev/sdb"
      clevis:
        tang:
        - url: https://tang.example.com
          thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
          advertisement: '{"payload": "...", "protected": "...", "signature": "..."}'
```
Much like any other render/conversion error, your pool will degrade: 
```
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-a79af855fcc5ad725c369aa16a2d6f32   True      False      False      3              3                   3                     0                      46m
worker   rendered-worker-8664798e8b1c6588299c00a261886237   True      False      True       3              3                   3                     0                      46m
```
```
    Last Transition Time:  2023-03-01T20:36:33Z
    Message:               Failed to render configuration for pool worker: translating Ignition config failed with error Invalid input config: tang offline provisioning is not supported in spec v3.3
    Reason:                
    Status:                True
    Type:                  RenderDegraded
```

And the render controller will whine about it: 
```
I0301 20:36:38.767708       1 render_controller.go:377] Error syncing machineconfigpool worker: translating Ignition config failed with error Invalid input config: tang offline provisioning is not supported in spec v3.3
```

For the record I'd absolutely love to have an admission controller to parse and reject it up front. Maybe someday :smile:

**Comment by jkyros** [comment]
Revendored now that https://github.com/coreos/ign-converter/pull/44 merged

File: pkg/controller/common/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkyros) [reply pkg/controller/common/helpers.go:552]: If we bump our default from 3.2 to 3.4 *without the MCO implementing the features described in 3.3 and 3.4* then yes we should do exactly what you describe.

**Comment by bgilbert** [comment]
> The MCO has traditionally tried to support an ignition version at least one release before it's actually used so we have the ability to roll back an upgrade, etc.

For the record, this is actually a new practice, driven by the [requirement for clean downgrade tests](https://bugzilla.redhat.com/show_bug.cgi?id=1947477).

> Added ignition 3.3 version support (but not support for the kargs feature yet)

The general design here is that 4.13 will support any configs declaring themselves as 3.4.0 that are also translatable to 3.2.0, right?  And then in 4.14, we can start enabling 3.4.0 features, which is fine because we only need to support downgrades to 4.13 when no 4.14-specific features are configured.

That seems okay in general, but if MCO 4.14 starts translating MachineConfig `kernelArguments` to a rendered config that uses Ignition `kernelArguments`, then MCO 4.14 will convert a MachineConfig without 4.14-specific features into a rendered config that 4.13 won't accept.

> Added ignition 3.4 version support (but not support for the offline tang feature yet)

Offline Tang shouldn't need explicit MCO support AFAIK; the field just needs to be passed through to Ignition.

However, I noticed that `reconcilable()` doesn't compare `Storage.Luks` in the old and new configs, so any MachineConfig changes that affect LUKS will be silently allowed and ignored for existing nodes.  This isn't a new problem, but might be worth fixing while here?

File: pkg/controller/common/helpers.go [hunk 2]
```diff
@@ -281,6 +284,50 @@ func ConvertRawExtIgnitionToV3(inRawExtIgn *runtime.RawExtension) (runtime.RawEx
 	return outRawExt, nil
 }
 
+// ConvertRawExtIgnitionToV3_4 ensures that the Ignition config in
+// the RawExtension is spec v3.4, or translates to it.
+func ConvertRawExtIgnitionToV3_4(inRawExtIgn *runtime.RawExtension) (runtime.RawExtension, error) {
+	// TODO(jkyros): since 3.4 is "ahead" of our current default 3.2, we're going "up" not down, which is
```
Comment (yuqi-zhang) [inline pkg/controller/common/helpers.go:290]: Nit: I think this comment was for the V3_3 conversion (`Once we bump to 3.3 this will be briefly obsolete` should be 3.4, and the rest will be future proofing)

File: pkg/controller/common/helpers.go [hunk 3]
```diff
@@ -495,39 +542,42 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
 // IgnParseWrapper parses rawIgn for both V2 and V3 ignition configs and returns
 // a V2 or V3 Config or an error. This wrapper is necessary since V2 and V3 use different parsers.
 func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
-	ignCfgV3_2, rptV3_2, errV3_2 := ign3.Parse(rawIgn)
-	if errV3_2 == nil && !rptV3_2.IsFatal() {
+	// ParseCompatibleVersion will parse any config <= N to version N
+	ignCfgV3, rptV3, errV3 := ign3_4.ParseCompatibleVersion(rawIgn)
```
Comment (yuqi-zhang) [inline pkg/controller/common/helpers.go:546]: Oh nice! We no longer have to chain ourselves :+1:

File: pkg/server/api.go [hunk 4]
```diff
@@ -136,7 +136,26 @@ func (sh *APIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		converted34, err := ctrlcommon.ConvertRawExtIgnitionToV3_4(conf)
+		if err != nil {
+			w.Header().Set("Content-Length", "0")
+			w.WriteHeader(http.StatusInternalServerError)
+			glog.Errorf("couldn't convert config for req: %v, error: %v", cr, err)
+			return
+		}
+		serveConf = &converted34
+
+	} else if reqConfigVer.Equal(*semver.New("3.3.0")) {
```
Comment (yuqi-zhang) [inline pkg/server/api.go:149]: I think RHCOS has been on 3.3 for awhile, and technically we've been feeding it 3.2 up to now. I don't think this would change any actual functionality though right?

**Review by yuqi-zhang — COMMENTED** [review]
I think this makes sense, thanks for working on this! Some questions below

**Comment by yuqi-zhang** [comment]
> However, I noticed that reconcilable() doesn't compare Storage.Luks in the old and new configs, so any MachineConfig changes that affect LUKS will be silently allowed and ignored for existing nodes. This isn't a new problem, but might be worth fixing while here?

I believe this is https://bugzilla.redhat.com/show_bug.cgi?id=1964701 + https://issues.redhat.com/browse/RFE-1894 ?

File: pkg/controller/common/helpers.go (hunk 2 above)
Comment (jkyros) [reply pkg/controller/common/helpers.go:290]: You correct sir, good catch :smile:

File: pkg/server/api.go (hunk 4 above)
Comment (jkyros) [reply pkg/server/api.go:149]: I didn't know that and wasn't sure if things cared if they got a different version than they asked for. 

It shouldn't change any actual functionality, but we can totally keep feeding them `3.2` if you're worried. 

That probably turns into "If they ask for a 3.x version  > our default, give them our default" .

**Comment by bgilbert** [comment]
> I believe this is https://bugzilla.redhat.com/show_bug.cgi?id=1964701 + https://issues.redhat.com/browse/RFE-1894 ?

IMO those issues are a bit different, in that they cover one case where the MCO might want to learn to reconcile `Luks` section changes.  For now, though, any and all changes to the `Luks` section are being silently ignored for existing nodes, rather than degrading the node as happens for other irreconcilable config changes.

File: pkg/server/api.go (hunk 4 above)
Comment (bgilbert) [reply pkg/server/api.go:149]: It should be fine to send a newer config version.  Ignition uptranslates older configs internally anyway.

**Comment by jkyros** [comment]
> IMO those issues are a bit different, in that they cover one case where the MCO might want to learn to reconcile Luks section changes. For now, though, any and all changes to the Luks section are being silently ignored for existing nodes, rather than degrading the node as happens for other irreconcilable config changes.

We were kind of "too chicken" to fix it by just degrading because the folks in those issues Jerry linked were fixing it by updating it in MachineConfig (for any new nodes that come online), and then retrofitting their old ones via SSH or whatever.  If we degrade, they'd have to retrofit out-of-band forever :/

**Comment by yuqi-zhang** [comment]
Right. I was more so pointing to the fact that some users may be depending on one behaviour already. If we want to switch to "degrade" or "MCO manages it" instead of "the silent treatment", I think we would need to consider the ramifications to switch. I would be tentatively be ok with us degrading, if only to warn the user changes may be coming in future versions

**Comment by jkyros** [comment]
> For the record, this is actually a new practice, driven by the [requirement for clean downgrade tests](https://bugzilla.redhat.com/show_bug.cgi?id=1947477).

That is...before the @jkyros era (so for me, it is a tradition!). I do very much appreciate the background though, I didn't know where it came from.  :smile: 

> The general design here is that 4.13 will support any configs declaring themselves as 3.4.0 that are also translatable to 3.2.0, right? And then in 4.14, we can start enabling 3.4.0 features, which is fine because we only need to support downgrades to 4.13 when no 4.14-specific features are configured.

Right, exactly. 

> That seems okay in general, but if MCO 4.14 starts translating MachineConfig `kernelArguments` to a rendered config that uses Ignition `kernelArguments`, then MCO 4.14 will convert a MachineConfig without 4.14-specific features into a rendered config that 4.13 won't accept.

And yep,  this right here is the "miserable" part. The moment those fields get used, we can't roll back anymore unless we filter that stuff out (which I think is okay to filter kargs, but not for tang). And the MCO would be syncing those fields automatically rather than someone "choosing to use a new feature". 

> Offline Tang shouldn't need explicit MCO support AFAIK; the field just needs to be passed through to Ignition.

I'm going to be really honest, I didn't know how much work the MCO was going to have to do for offline tang, I just knew the rotation of the tang keys was a little more involved, I assumed there might be something the MCO had to do here, and was allowing for us having to do work to reconcile but wasn't sure. 

> I would be tentatively be ok with us degrading, if only to warn the user changes may be coming in future versions

The tang rotation was on the list for 4.14, I think we have enough context to do it, I'd personally rather finally fix it for real :smile: 

So then this here PR probably either goes: 

1. We keep it on 3.2 and kargs either waits until 4.15 or we backport something
2. We keep it on 3.2 and put some code in here to remove the kargs from > 3.2 so they can't blow us up in a rollback (mco and ignition kargs should match, right?) 
3. We bump default to 3.4 now and just silently ignore kargs/offline tang (seems unkind, but sure) 
4. We bump default to 3.4 but whine in the logs that "these 3.3/3.4 fields were supplied but are unimplemented by the MCO and the MCO is ignoring them"

**Comment by bgilbert** [comment]
> I'm going to be really honest, I didn't know how much work the MCO was going to have to do for offline tang, I just knew the rotation of the tang keys was a little more involved, I assumed there might be something the MCO had to do here, and was allowing for us having to do work to reconcile but wasn't sure.

Ah yeah, it does have some impact on key rotation, but hopefully not much.  I think it should mostly mean that there's one more field to pass through to the `clevis` command when performing the rotation.

> 2. We keep it on 3.2 and put some code in here to remove the kargs from > 3.2 when we downtranslate so they can't blow us up in a rollback (mco and ignition kargs should match, right?)

I'd tend to favor this approach, though I'm not sure of the details.  If "new-style" kargs support ends up _dropping_ the MC kargs field from the rendered config (would it?) then the 4.13 MCO would need to move Ignition's `kernelArguments.shouldExist` field back into the MachineConfig `kernelArguments` field.  And I don't think there's an existing equivalent to Ignition's `kernelArguments.shouldNotExist`, so we'd presumably want to error if present.

> 3. We bump default to 3.4 now and just silently ignore kargs/offline tang (seems unkind, but sure)

If this includes the rendered configs, that would violate the downgrade constraint, right?

**Comment by jkyros** [comment]
I agree, I think "Option 2."  is what we want. I was kind of hoping we could postpone the finer points of the kargs migration until later, but I guess we have to think through at last some of it now :smile:  

I rebased and added some code here to handle the downgrade in "Option 2." fashion.  I _think_ this enough to safely handle it regardless of what we decide next version, but I might be missing something. 

What it does is: 
- Strips out the kargs during down-translation, which should be okay because everything uses IgnParseWrapper or ParseAndConvertConfig
- During render/merge it merges any of the kargs in ignition's ShouldExist to MachineConfig's kargs ( so it shouldn't matter if we leave MCO kargs populated or not in the next version) 
- If ShouldNotExist is populated, we degrade during render (which should be fine because the MCO won't be populating that automatically in the next version, regardless of what we do)

And this works (the args get "migrated" to MachineConfig in the rendered-config and applied): 
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name:  99-ign-kargs 
spec:
  config:
   ignition:
     version: 3.3.0
   kernelArguments:
     shouldExist:
     - foobar
     - baz boo
```

And the pool degrades on this as expected: 
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name:  99-ign-kargs 
spec:
  config:
   ignition:
     version: 3.3.0
   kernelArguments:
     shouldExist:
     - foobar
     - baz boo
     shouldNotExist:
     - raboof
```
And...I don't think I broke anything else, but we'll see.

**Review by yuqi-zhang — APPROVED** [review]
I think this should cover us for enough cases and warrants getting into 4.13 to make future transitions easier. If needed, we should merge this as a bug, but leaving my approval and letting others take a final look as well

File: pkg/controller/common/helpers.go [hunk 5]
```diff
@@ -539,12 +551,44 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
...
+// function is used by the render controller to extract the kargs from ignition and migrate them to
+// MachineConfig so newer versions that specify kargs in ignition will still work during a downgrade.
+// TODO(jkyros): remove this when we raise the ignition default to 3.4
+func ExtractIgnitionKargsFor4_13(rawIgn []byte) ([]ign3_4types.KernelArgument, error) {
+	ignCfgV3, rptV3, errV3 := ign3_4.ParseCompatibleVersion(rawIgn)
+	// No kargs in an empty config
+	if errors.Is(errV3, ign3errors.ErrEmpty) {
+		return nil, nil
+	}
+	if errV3 == nil && !rptV3.IsFatal() {
```
Comment (bgilbert) [inline pkg/controller/common/helpers.go:565]: `errV3` shouldn't be nil if the report is fatal.  (And if it were, we'd silently return `nil, nil` from this function.)

File: pkg/controller/common/helpers.go [hunk 6]
```diff
@@ -539,12 +551,44 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
...
+		// We can't reconcile ShoultNotExist
```
Comment (bgilbert) [inline pkg/controller/common/helpers.go:566]:
```suggestion
		// We can't reconcile ShouldNotExist
```

File: pkg/controller/common/helpers.go [hunk 7]
```diff
@@ -539,12 +551,44 @@ func ValidateMachineConfig(cfg mcfgv1.MachineConfigSpec) error {
...
+		if len(ignCfgV3.KernelArguments.ShouldNotExist) > 0 {
+			return nil, fmt.Errorf("Ignition KernelArguments.ShouldNotExist is not supported in this release ( ShouldNotExist: %s was supplied )", ignCfgV3.KernelArguments.ShouldNotExist)
+		}
+
+		// But we can stuff ShouldExist in MachineConfig, so send those back
```
Comment (bgilbert) [inline pkg/controller/common/helpers.go:571]: One note: Ignition `KernelArguments` are allowed to contain spaces, and this is encouraged when multiple arguments are order-dependent.  For example, `console=tty0 console=ttyS0,115200` might be written as a single KernelArgument.  The MCO karg code appears to always split on spaces, though.  Here we're only concerned with downgrades, so if the future MCO never creates these compound arguments when converting from MachineConfig, this should all be fine.

**Review by bgilbert — COMMENTED** [review]
I'm not very familiar with this codebase, but LGTM.

**Comment by jkyros** [comment]
Rebased and added review fixes.

**Comment by sergiordlr** [comment]
Hello! @jkyros @rioliu-rh 

In the previous comment we describe what we did in order to verify this PR.

We found mainly 2 problems:

- The MC with the luks.tang configuration did not make the MCP degraded, but since the comments in this PR it should have been degraded.

- MCs deploying files using ignition version 2.2.0 degrade the MCP. This is a regression.

Unless those problems are expected and intended (specially  the second problem, which is a regression) we can't add the the approved label.

**Comment by jkyros** [comment]
@sergiordlr I always appreciate you checking, excellent catches. 

> MCs deploying files using ignition version 2.2.0 degrade the MCP. This is a regression.

I was indeed a clown, and forgot about ignition 2 going through that kernel arg extraction function. I've adjusted it so it will allow it to pass through without generating an error.  We also didn't have any test cases that stuffed an igniton 2 version through MergeMachineConfig to catch that -- I added one so I can't silently break it again. :smile:

> The MC with the luks.tang configuration did not make the MCP degraded, but since the comments in this PR it should have been degraded.

This...I think is worse than I initially thought. Glad you tried it. I was initially like "oh his Tang yaml was missing the storage section, so it just got ignored" so then I fixed that and tried:
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name:  99-cant-convert-34
spec:
  config:
    ignition:
      version: 3.4.0
    storage:
      luks:
      - name: luks-tang
        device: "/dev/sdb"
        clevis:
          tang:
          - url: https://tang.example.com
            thumbprint: REPLACE-THIS-WITH-YOUR-TANG-THUMBPRINT
            advertisement: '{"payload": "...", "protected": "...", "signature": "..."}'
```



and got apparently the same behavior you described. I was confused as to why the MCP wouldn't be degrading and it's...because the controller doesn't survive long enough to make it degrade once you apply a luks config: 
```
	/go/src/github.com/openshift/machine-config-operator/vendor/github.com/coreos/ignition/v2/config/translate/translate.go:138 +0x276
github.com/coreos/ignition/v2/config/translate.translator.Translate({{0xc000e774a0?, 0x1e851c0?, 0x2404d88?}}, {0x208e700?, 0xc00095f5a8?}, {0x208e400?, 0xc00150d798?})
	/go/src/github.com/openshift/machine-config-operator/vendor/github.com/coreos/ignition/v2/config/translate/translate.go:186 +0x205
github.com/coreos/ign-converter/translate/v33tov32.translateConfig({{{{0x0, 0x0, 0x0}, {0x0, {...}, 0x0, {...}}}, {0x0, 0x0, {0x0, ...}}, ...}, ...})
	/go/src/github.com/openshift/machine-config-operator/vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go:94 +0x1b6
github.com/coreos/ign-converter/translate/v33tov32.Translate({{{{0x0, 0x0, 0x0}, {0x0, {...}, 0x0, {...}}}, {0x0, 0x0, {0x0, ...}}, ...}, ...})
	/go/src/github.com/openshift/machine-config-operator/vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go:112 +0x19d
github.com/openshift/machine-config-operator/pkg/controller/common.IgnParseWrapper({0xc0003bc140, 0x13c, 0x140})
...
panic: Translate needs to be called on pointers [recovered]
	panic: Translate needs to be called on pointers
```

The ignition translator panics on the 3.3 -> 3.2 conversion and kills the `machine-config-controller` if `Clevis` is present. A succinct reproduction is probably something like [this](https://gist.github.com/jkyros/8d481da868192197d714e723ce35f6d0). It looks like it has to do with fields switching whether they were pointers or not between ignition versions -- I'll add a comment after this one with some code refs. 

In summary: 

- I believe I've fixed the ignition 2.2 issue (whoops)
- I think we're going to have to fix the "Clevis downtranslation to 3.2" issue in the ignition translator, but I hacked a fix in here in `vendor/` to illustrate the issue.  
- I'm a little spooked that at some point the MCO needs more ignition test cases and/or we should probably catch panics during ignition parses

File: vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go [hunk 8]
```diff
@@ -47,7 +48,13 @@ func translateRaid(old old_types.Raid) (ret types.Raid) {
...
-	tr.Translate(old.Clevis, &ret.Clevis)
+
+	// this goes from "not pointer" to "pointer" so we need to populate it
+	// if old.Cievis isn't empty and we have something to translate in
+	if !reflect.DeepEqual(old.Clevis, old_types.Clevis{}) {
+		ret.Clevis = &types.Clevis{}
+		tr.Translate(&old.Clevis, ret.Clevis)
+	}
 	tr.Translate(&old.Device, &ret.Device)
 	tr.Translate(&old.KeyFile, &ret.KeyFile)
```
Comment (jkyros) [inline vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: @bgilbert @prestist  I hit kind of a snag here with how Clevis gets down-translated from 3.3 to 3.2, seeing that some of the 3.2 fields are pointers but those same fields in 3.3 are actual structs. (i was initially getting "Translate needs to be called on pointers" panics but when I adjusted it to account for just that, I was getting a panic anyway because the "target" pointer in the 3.2 structure was nil by default) 

I tried to work around it like this in...a way that seemed reasonable after poking around the rest of the translator for precedent, but on the whole I'm still probably ignorant of the rest of your code base and general practices/philosophy. 

Just looking for some guidance -- I can PR something like this to the converter if a fix along these lines is acceptable, or you can tell me how I'm using it wrong, etc :smile:

Comment (prestist) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: I am taking a quick look now :)

Comment (prestist) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: @jkyros could we get a pr upstream with this so you can just revendor?

Comment (jkyros) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: Yep, thanks! Really appreciate you taking a look, I've opened a PR: https://github.com/coreos/ign-converter/pull/47

Comment (bgilbert) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: The converter doesn't really have practices or philosophy; it's pretty hacky.  :slightly_smiling_face:

Comment (bgilbert) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: https://github.com/coreos/ign-converter/pull/47 merged!

Comment (jkyros) [reply vendor/github.com/coreos/ign-converter/translate/v33tov32/v33tov32.go]: Thank you both for your time on this, I very much appreciate it! :smile:

**Comment by jkyros** [comment]
Alright, I've re-vendored the ignition converter and rebased.

@sergiordlr this fixes the panic so this should be ready for another look when you have a moment. Thank you for your engagement and your patience.

**Comment by sergiordlr** [comment]
Verified using IPI on GCP with proxy.

Steps that we followed in order to verity the PR (similar to those in https://github.com/openshift/machine-config-operator/pull/3576#issuecomment-1479750728):

- **Create a MC using ignition version 3.4  and "shouldNotExist" attribute**

The behavior was the expected one.The result was the same that we got in the first verification.

- **Create a MC using ignition version 3.4 with kernel arguments**

The behavior was the expected one. The result was the same that we got in the first verification.

- **Check that there is no problem when we ask MCS for different ignition versions:**

The behavior was the expected one. The result was the same that we got in the first verification. 

Note that the kernel arguments are not present in the 3.4.0/3.3.0 versions of the ignition configs.

```
# curl -H "Accept: application/vnd.coreos.ignition+json; version=3.4.0" -k https://api.sregidor-ver3.qe.gcp.devcluster.openshift.com:22623/config/worker | jq  | grep enforce

```

- **Create a MC using unsupported luks.tang configuration:**

The MCP was not degraded. Since this comment this issue is expected https://github.com/openshift/machine-config-operator/pull/3576#issuecomment-1491114911

We can think about creating a different jira issue to track this problem.


- **Create a MC using kernels arguments and 3.2.0 version.**

The behavior was the expected one. The result was the same that we got in the first verification. 

- **Create a MC using both ways to declare kernel arguments**

The behavior was the expected one. The result was the same that we got in the first verification. 

- **Create a random test file in /etc  using MCs with ignition versions 2.2.0, 3.1.0, 3.2.0, 3.3.0, 3.4.0**

All files were created properly. The error in ignition version 2.2.0 does not happen anymore.

- **Create a random unit using MCs with ignition versions 2.2.0 and 3.4.0**

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: mc-unit-22
spec:
  config:
    ignition:
      version: 2.2.0
    systemd:
      units:
      - name: my22unit.service
        contents: |
          [Unit]
          Description=A simple oneshot service
          
          [Service]
          Type=oneshot
          ExecStart=/bin/bash -c "echo Hello world"
          RemainAfterExit=yes

          [Install]
          WantedBy=multi-user.target
```

Both 2.2.0 and 3.4.0 version behaved as expected. New rendered 3.2.0 machineconfigs were created and the units worked as expected

```
sh-5.1# systemctl status my22unit.service
○ my22unit.service - A simple oneshot service

sh-5.1# systemctl status my34unit.service
○ my34unit.service - A simple oneshot service

```


- **All Critical/Medium/Low priority automated e2e test cases were executed using a IPI GCP cluster with proxy**

No issues found


- **All High priority automated e2e test cases were executed using a IPI on GCP cluster without proxy**

One issue found. We have a test (TC OCP-42390) to make sure that when the ignition version is empty,  the MCP is degraded showing this message:

```
parsing Ignition config failed: unknown version. Supported spec versions: 2.2, 3.0, 3.1, 3.2
```

Instead of this message now we get this other message

```
Failed to render configuration for pool worker: parsing Ignition config spec v3 failed with error: invalid config version (couldn't parse)
```

We can modify the expected test message, but we would like to make sure that the message that we use is the one that we actually want to use.

I think that the TC's expected message is more descriptive and helpful than the one we are reporting now. 

Apart from that minor issue, no failure was found and everything is working as expected. Once we decide if we want to modify this error message or keep it, we will add the qe-approved label.

Thank you very much for your great job!!

**Comment by jkyros** [comment]
Rebased and added a more helpful error message for "Invalid Version" . 

@sergiordlr I kind of "split the difference" on the error message. The new ignition parsing introduced in this PR differentiates between `ErrInvalidVersion` ("I can't parse this into a version number that makes sense") and `ErrUnknownVersion` ("I can parse this into a version number, but it's not a version number I understand"). I wanted to keep that differentiation, but I also wanted to be helpful so the user knows what to put in there. 

The new error message is: 
> parsing Ignition config failed: invalid version. Supported spec versions: 2.2, 3.0, 3.1, 3.2, 3.3, 3.4 

(I just handle that error type explicitly as a separate case, and changed unknown to invalid, so we could tell by the logs which case we hit)

**Comment by sergiordlr** [comment]
The error reported now is similar to the previous one.

We  add the qe-approved label. @jkyros Thank you very much!!

/label qe-approved

**Review by sinnykumari — COMMENTED** [review]
Overall LGTM, nice work John!
/lgtm

**Comment by jkyros** [comment]
Thanks for looking at it Sinny, and since we have QE approval, it looks like we're good to go 
/hold cancel

**Comment by jkyros** [comment]
What a weird e2e-aws-ovn failure. That's been green the whole time. I'm struggling to correlate anything we've done here to service recreation failing. 

`{"ProwJobName":"pull-ci-openshift-machine-config-operator-master-e2e-aws-ovn","ProwJobRunID":1658595958570618880,"Release":"Presubmits","CompareRelease":"4.14","Tests":[{"Name":"[sig-network] Services should work after the service has been recreated [Suite:openshift/conformance/parallel] [Suite:k8s]","Risk":{"Level":{"Name":"Medium","Level":50},"Reasons":["This test has passed 97.59% of 83 runs on release 4.14 [amd64 aws ha ovn] in the last week."]},"OpenBugs":[]}],"OverallRisk":{"Level":{"Name":"Medium","Level":50},"Reasons":["Maximum failed test risk: Medium"]},"OpenBugs":[]}`

/test e2e-aws-ovn

**Comment by jkyros** [comment]
Looks like we hit a grace period on that one. Feels like global issues. We'll try one more time. 

`{"component":"entrypoint","file":"k8s.io/test-infra/prow/entrypoint/run.go:254","func":"k8s.io/test-infra/prow/entrypoint.gracefullyTerminate","level":"error","msg":"Process did not exit before 10m0s grace period","severity":"error","time":"2023-05-17T17:06:08Z"}
{"component":"entrypoint","error":"os: process already finished","file":"k8s.io/test-infra/prow/entrypoint/run.go:256","func":"k8s.io/test-infra/prow/entrypoint.gracefullyTerminate","level":"error","msg":"Could not kill process after grace period","severity":"error","time":"2023-05-17T17:06:08Z"}
{"component":"entrypoint","error":"process aborted","file":"k8s.io/test-infra/prow/entrypoint/run.go:79","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2023-05-17T17:06:08Z"}
error: failed to execute wrapped command: exit status 130`

/test e2e-aws-ovn
